code/datums/components/leanable.dm

```diff
@@ -0,0 +1,146 @@
+		stopped_leaning(src)
+
+/datum/component/leanable/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_MOUSEDROPPED_ONTO)
+	UnregisterSignal(leaning_mob, COMSIG_LIVING_STOPPED_LEANING)
+	UnregisterSignal(parent, COMSIG_ATOM_DENSITY_CHANGED)
+	leaning_mob = null
+
+/datum/component/leanable/proc/stopped_leaning(obj/source)
+	SIGNAL_HANDLER
+	qdel(src)
+
+/datum/component/leanable/proc/mousedrop_receive(atom/source, atom/movable/dropped, mob/user, params)
+	if(dropped != user) //Have we been dropped on a valid leanable object?
+		return FALSE
+	var/mob/living/leaner = dropped
+	leaning_mob = leaner
+	if(!iscarbon(dropped) && !iscyborg(dropped)) //Are we not a cyborg or carbon?
+		return FALSE
+	if(!(usr == leaner)) //Are we trying to lean someone else?
+		return FALSE
+	if(leaner.incapacitated(IGNORE_RESTRAINTS) || leaner.stat != CONSCIOUS || leaner.notransform || leaner.buckled || leaner.body_position == LYING_DOWN) //Are we in a valid state?
+		return FALSE
+	if(HAS_TRAIT_FROM(leaner, TRAIT_UNDENSE, TRAIT_LEANING)) //Are we leaning already?
+		return FALSE
+	if(ISDIAGONALDIR(get_dir(leaner, source))) //Not leaning on a corner, idiot
+		return FALSE
+	if(!is_currently_leanable) //Is the object currently able to be leaned on?
+		return FALSE
+
+	leaner.apply_status_effect(STATUS_EFFECT_LEANING, source, leaning_offset)
+	return TRUE
+
+/datum/component/leanable/proc/on_density_change()
+	SIGNAL_HANDLER
+	is_currently_leanable = !is_currently_leanable
+	if(!is_currently_leanable)
+		leaning_mob.stop_leaning(src)
+
+/**
+ * Makes the mob lean on an atom
+ * Arguments
+ *
+ * * atom/lean_target - the target the mob is trying to lean on
+ * * leaning_offset - pixel offset to apply on the mob when leaning
+ */
+/mob/living/proc/start_leaning(atom/lean_target, leaning_offset)
+	var/new_x = lean_target.pixel_x + base_pixel_x + body_position_pixel_x_offset
+	var/new_y = lean_target.pixel_y + base_pixel_y + body_position_pixel_y_offset
+	switch(get_dir(src, lean_target))
+		if(SOUTH)
+			new_y -= leaning_offset
+		if(NORTH)
+			new_y += leaning_offset
+		if(WEST)
+			new_x -= leaning_offset
+		if(EAST)
+			new_x += leaning_offset
+
+	animate(src, 0.2 SECONDS, pixel_x = new_x, pixel_y = new_y)
+	ADD_TRAIT(src, TRAIT_UNDENSE, TRAIT_LEANING)
+	visible_message(
+		span_notice("[src] leans against [lean_target]."),
+		span_notice("You lean against [lean_target]."),
+	)
+	leaned_object = lean_target
+	RegisterSignals(src, list(
+		COMSIG_MOB_CLIENT_MOVED,
+		COMSIG_HUMAN_DISARM_HIT,
+		COMSIG_MOVABLE_PULLED,
+		COMSIG_PARENT_QDELETING,
+		COMSIG_LIVING_RESIST, // Pressing B!
+		COMSIG_LIVING_MINOR_SHOCK, //If we are hit by a stunbaton
+		COMSIG_LIVING_RESTING_UPDATED //If we are downed
+	), PROC_REF(stop_leaning))
+	RegisterSignal(src, COMSIG_ATOM_TELEPORT_ACT, PROC_REF(teleport_away_while_leaning))
+	RegisterSignal(lean_target, COMSIG_AIRLOCK_OPEN, PROC_REF(airlock_opened))
+	RegisterSignal(lean_target, COMSIG_VEHICLE_MOVE, PROC_REF(mech_moved))
```

Comment:
What if the target is a movable that moves, but isn't a vehicle or airlock?

For example:
- Lean on a non-anchored window
- Have another mob move the non-anchored window